source/dataman/CLS/CLSDbUpgrade1Pt2.cpp

```diff
@@ -0,0 +1,133 @@
+#include "CLSDbUpgrade1Pt2.h"
+
+#include <QStringBuilder>
+
+#include "dataman/database/AMDbObjectSupport.h"
+#include "util/AMErrorMonitor.h"
+
+CLSDbUpgrade1Pt2::CLSDbUpgrade1Pt2(const QString &facilityName, const QString &databaseNameToUpgrade, QObject *parent)
+	: AMDbUpgrade(databaseNameToUpgrade, parent)
+{
+	targetFacilityName_ = facilityName;
+}
+
+CLSDbUpgrade1Pt2::~CLSDbUpgrade1Pt2()
+{
+
+}
+
+QStringList CLSDbUpgrade1Pt2::upgradeFromTags() const
+{
+	return QStringList();
+}
+
+bool CLSDbUpgrade1Pt2::upgradeNecessary() const
+{
+	// run update when:
+	//
+	//   - The AMDbObjectThumbnails_table doesn't have thumbnail for the given beamline name or has more than one thumbnails for AMFacilityTable
+
+	QString thumbnailTableName = AMDbObjectSupport::thumbnailTableName();
+	QMap<int, QString> facilityTableThumbnails;
+
+	QSqlQuery thumbnailTableQuery = databaseToUpgrade_->select(thumbnailTableName, "id, title", QString("objectTableName=%1").arg("AMFacility_table") );
+	if (thumbnailTableQuery.exec()) {
+		while (thumbnailTableQuery.next()) {
+			facilityTableThumbnails.insert(thumbnailTableQuery.value(0).toInt(), thumbnailTableQuery.value(1).toString()); //id, title
+//			qDebug() << "==== " << thumbnailTableQuery.value(0).toInt() << thumbnailTableQuery.value(1).toString();
+		}
+	}
+	thumbnailTableQuery.finish();
+
+	bool correctThumbnailTableData = (facilityTableThumbnails.count() == 1) && (facilityTableThumbnails.values().at(0) == targetFacilityName_);
+//	qDebug() << "==== " << correctThumbnailTableData;
+
+	return !correctThumbnailTableData;
+}
+
+bool CLSDbUpgrade1Pt2::upgradeImplementation()
+{
+	int dbResult = -1;
+
+	// Prologue.
+	////////////////////////////////////////////////
+
+	databaseToUpgrade_->startTransaction();
+
+	QString thumbnailTableName = AMDbObjectSupport::thumbnailTableName();
+
+	// Stage 1: update the thumbnail table if required
+	//   - The AMDbObjectThumbnails_table doesn't have thumbnail for the given beamline name
+	QMap<int, QString> facilityTableThumbnails;
+	QSqlQuery thumbnailTableQuery = databaseToUpgrade_->select(thumbnailTableName, "id, title", QString("objectTableName=%1").arg("AMFacility_table") );
+	if (thumbnailTableQuery.exec()) {
+		while (thumbnailTableQuery.next()) {
+			facilityTableThumbnails.insert(thumbnailTableQuery.value(0).toInt(), thumbnailTableQuery.value(1).toString()); //id, title
+		}
+	}
+	thumbnailTableQuery.finish();
+
+	bool correctThumbnailTableData = (facilityTableThumbnails.count() == 1) && (facilityTableThumbnails.values().at(0) == targetFacilityName_);
```

Comment:
We are expecting only one for the facility table thumbnail, and the the first value should match with the target name